[STAGE-6.1] Additional Tasks

polinatonko opened the pull request:
Complete below steps:
1. Add CHANGELOG.md file with description of changes based on project version 
2. Update README.md with project description, stack and local setup
3. Add Application Autoscaler for application (cpuutil metric)
4. Configure sending emails with course reminders asynchronously (Async annotation and custom thread pool)

polinatonko committed:
feat: add autoscaler service

polinatonko committed:
feat: send course notifications asynchronously

polinatonko committed:
docs: edit README.md, add CHANGELOG.md

polinatonko committed:
fix: update API URIs in integration tests, test profile configuration properties

polinatonko committed:
docs: update CHANGELOG.md

polinatonko committed:
docs: update README.md

polinatonko committed:
docs: update README.md

polinatonko committed:
fix: autoscaler binding in mta.yaml

polinatonko committed:
Merge branch 'final-stage-5' into stage-7

polinatonko committed:
chore: configure json logs format for cloud

polinatonko committed:
chore: update logback configuration and application properties

polinatonko committed:
fix: security configuration, open api docs

polinatonko committed:
Merge branch 'stage-5' into stage-7

polinatonko committed:
chore: update security configuration

polinatonko committed:
chore: update pom.xml

SN4NTR commented on btp/autoscaler/config.json:
```diff
@@ -0,0 +1,18 @@
+  "scaling_rules": [
+    {
+      "metric_type": "cpuutil",
+      "threshold": 85,
```
to be more precise, add `breach_duration_secs ` property to explicitly define how much time this condition must be met until amount of instances is increased: https://help.sap.com/docs/application-autoscaler/application-autoscaler/parameters-for-dynamic-scaling-policy?locale=en-US&q=breach_duration_secs

E.g.:
```
{
      "metric_type": "cpuutil",
      "operator": ">=",
      "threshold": 85,
      "breach_duration_secs": 60,
      "adjustment": "+1"
    },
    {
      "metric_type": "cpuutil",
      "operator": "<",
      "threshold": 85,
      "breach_duration_secs": 60,
      "adjustment": "-1"
    }
```

SN4NTR commented on src/main/java/org/example/learningsystem/core/security/authority/UserAuthority.java:
```diff
@@ -0,0 +1,35 @@
+ * Represents user authorities in the business domain.
+ */
+@AllArgsConstructor
+public enum UserAuthority {
```
I recommend to rename it to `UserRole`, because authority - is what exactly we can do, but role - is group of these authorities
```
Read, Write, Manage, Administrate - authorities
ADMIN, MANAGER - roles

And ADMIN role can contain Read, Write, Manage and Administrate authorities
```

In XSUAA we are talking in context of authorities (it doesn't use concept of roles), so it means that all scopes in JWT will represent authorities only (like Administrate, Manage, TenantProvisioning)
So in code we have to do check using hasAnyAuthority('Administrate') for example

SN4NTR commented on src/main/java/org/example/learningsystem/multitenancy/service/TenantService.java:
```diff
@@ -0,0 +1,28 @@
+
+    private final MultiTenantDataSource multiTenantDataSource;
+
+    public void executeForAllTenants(Runnable runnable) {
```
rename to executeForAll, as Tenants domain can be understood from class name

SN4NTR commented on src/main/resources/application.yaml:
```diff
@@ -21,6 +21,10 @@ spring:
     encoding: UTF-8
 
 basic-authentication:
+  admin:
```
these variables better include in local profile

polinatonko committed:
refactor: additional tasks

Edit xs-security.json and Spring security configuration, Autoscaler
policy, and profile properties.

SN4NTR commented on src/main/resources/application-cloud.yaml:
```diff
@@ -16,6 +16,11 @@ spring:
+basic-authentication:
+  manager:
+    username: manager@gmail.com
+    password: manager
```
and these properties should in user-provided service and read via environment variables :)

polinatonko committed:
chore: edit basic-authentication config in cloud
